twill-yarn/src/main/hadoop20/org/apache/twill/internal/yarn/Hadoop20YarnAppClient.java

```diff
@@ -176,22 +184,27 @@ private void addRMToken(ContainerLaunchContext context) {
 
   @Override
   public ProcessController<YarnApplicationReport> createProcessController(ApplicationId appId) {
-    return new ProcessControllerImpl(yarnClient, appId);
+    return new ProcessControllerImpl(createYarnClient(), appId);
   }
 
   @Override
   public List<NodeReport> getNodeReports() throws Exception {
-    return this.yarnClient.getNodeReports();
+    YarnClient yarnClient = createYarnClient();
+    try {
+      return yarnClient.getNodeReports();
+    } finally {
+      yarnClient.stop();
+    }
   }
 
   @Override
   protected void startUp() throws Exception {
-    yarnClient.start();
+    // no-op. We will create and start YarnClients, on demand
```

Comment:
Maybe we should make `YarnAppClient` not a `Service` at all. It seems like there is no need for any lifecycle management.